src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp

```diff
@@ -0,0 +1,222 @@
+    // underlying boolean value will represent whether the type can be returned
+    // or not.
+    //
+    // This works using a behavior called SFINAE. We define two overloads for
+    // CanBeReturned: one that inherits from true_type and one that inherits
+    // from false_type. When C++ is deciding which overload to choose, it
+    // starts with the true_type one, since it has the most specific signature.
+    // If the overload isn't ill-formed (i.e. the expression in decltype
+    // successfully compiles), the compiler will select that overload.
+    // Otherwise, it will move on to the false_type overload, which is never
+    // ill-formed. This means that the boolean value of the type is known at
+    // compile-time, and can be compared against an expected value.
+    //
+    // We use FromRemoteResult to test for whether the type can be returned
+    // because while calling BindResult on an un-returnable type would also
+    // fail to compile, the compiler can't detect compilation failures within
+    // that function. It can only reason about the flat expression in decltype.
+    template<class, class = std::void_t<>>
+    struct CanBeReturned : std::false_type{};
+
+    template<class T>
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    TEST_CLASS(UiaOperationAbstractionTests)
+    {
+    public:
+        // Asserts that you can get the name of a UiaElement.
+        void ElementGetNameTest(const bool useRemoteOperations)
+        {
+            ModernApp app(L"Microsoft.WindowsCalculator_8wekyb3d8bbwe!App");
+            app.Activate();
+            auto calc = WaitForElementFocus(L"Display is 0");
+
+            winrt::com_ptr<IUIAutomation> automation;
+            THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+            UiaOperationAbstraction::Initialize(useRemoteOperations, automation.get());
+            auto cleanup = wil::scope_exit([&]()
+            {
+                UiaOperationAbstraction::Cleanup();
+            });
+
+            auto scope = UiaOperationScope::StartNew();
+
+            UiaElement element = calc;
+            auto name = element.GetName(false /*useCachedApi*/);
+            scope.BindResult(name);
+
+            scope.Resolve();
+
+            Assert::AreEqual(std::wstring(static_cast<wil::shared_bstr>(name).get()), std::wstring(L"Display is 0"));
+        }
+        TEST_METHOD(ElementGetNameLocalTest)
+        {
+            ElementGetNameTest(false);
+        }
+
+        TEST_METHOD(ElementGetNameRemoteTest)
+        {
+            ElementGetNameTest(true);
+        }
+
+        // Asserts that GetEnclosingElement can be called with a cache request.
+        void GetEnclosingElementTest(const bool useRemoteOperations)
```

Comment:
>GetEnclosingElementTest [](start = 13, length = 23)

I would rename the test and test helper functions to reflect what they really test.

(I was about to write a comment in the last test in the file that it is important to test the case where `Text` pattern is available only to realize that this test does it :))

Reply:
True, they should probably mention cache requests in the name :). The reason why I have this test and the GetParentElement test is that these two functions handle cache requests slightly differently, and I wanted coverage across them. I could maybe also rename these tests to CacheRequestPatternMethodTest and CacheRequestNavigationMethodTest.